cmd/clusterctl/clusterdeployer/clusterdeployer.go

```diff
@@ -216,29 +215,9 @@ func (d *ClusterDeployer) saveProviderComponentsToCluster(factory provider.Compo
 }
 
 func deleteClusterAPIObjectsInAllNamespaces(client clusterclient.Client) error {
-	var errorList []string
-	klog.Infof("Deleting MachineDeployments in all namespaces")
-	if err := client.DeleteMachineDeployments(""); err != nil {
-		err = errors.Wrap(err, "error deleting MachineDeployments")
-		errorList = append(errorList, err.Error())
-	}
-	klog.Infof("Deleting MachineSets in all namespaces")
-	if err := client.DeleteMachineSets(""); err != nil {
-		err = errors.Wrap(err, "error deleting MachineSets")
-		errorList = append(errorList, err.Error())
-	}
-	klog.Infof("Deleting Machines in all namespaces")
-	if err := client.DeleteMachines(""); err != nil {
-		err = errors.Wrap(err, "error deleting Machines")
-		errorList = append(errorList, err.Error())
-	}
 	klog.Infof("Deleting Clusters in all namespaces")
 	if err := client.DeleteClusters(""); err != nil {
-		err = errors.Wrap(err, "error deleting Clusters")
-		errorList = append(errorList, err.Error())
-	}
-	if len(errorList) > 0 {
-		return errors.Errorf("error(s) encountered deleting objects from bootstrap cluster: [%v]", strings.Join(errorList, ", "))
+		return errors.Errorf("error encountered deleting objects from bootstrap cluster: [error deleting Clusters: %v]", err)
```

Comment:
Will this run into any finalizer-related issues? I'm not overly familiar with the v1alpha2 related flow, but in v1alpha1 we removed the finalizers prior to delete to ensure that deletion wasn't blocked by the finalizer and that if the controllers were accidentally brought back up on the bootstrapping machine it wouldn't attempt to reconcile the resources that were already moved.

Reply:
Oh, are we using this function to go in both directions (create: bootstrap -> workload, delete: workload -> bootstrap)?

Reply:
We should probably block and wait for the cluster object to be deleted from the server, (unless client.DeleteClusters already does so). That should be enough to guarantee that the object tree has been deleted, in v1a2 the Cluster issues deletion requests to all of its children.

Reply:
Looking at the code, I think this function is only used by `clusterctl delete`. Pivot uses different functions that do remove the finalizers (ForceDeleteXYZ)

Reply:
@vincepri I'm pretty sure we added the same foreground deletion logic to v1alpha1, didn't we?

Reply:
I think so, but only after a certain version

Reply:
DeleteAllOf was doing `metav1.DeletionPropagationOrphan` so I changed it to `metav1.DeletionPropagationForeground` so the object tree will be deleted and the call will block.

Reply:
When did that change go in? I did not retest after rebasing. I was previously sync'd to v0.2.3. In my testing the call to DeleteAllOf returned immediately. Looking at the code, the default behavior of the client is `metav1.DeletePropagationOrphan`.

Reply:
Prior to releasing v0.2.0 (https://github.com/kubernetes-sigs/cluster-api/pull/1225).

Yes, it will return immediately - you'd have to add code to wait for it. When you run `kubectl delete cluster/foo`, it uses `DeletePropagatonBackground` and then waits for the cluster to be removed from the apiserver. It's because of kubectl's use of background propagation that we had to do what's in #1225.

Reply:
@ncdc PTAL at my approach to waiting for delete to complete.

Reply:
I took the approach by adding a method called `DeleteAllClusters` to the Cluster `Client` interface. I added a method that would just iterate through all the namespaces for clusters and delete each of them. 
https://github.com/wfernandes/cluster-api/blob/c644d0219dd970aebc27e0eb044c86481307b595/cmd/clusterctl/clusterdeployer/clusterclient/clusterclient.go#L540-L561

This could then be called by the [`clusterdeployer.go`](https://github.com/wfernandes/cluster-api/blob/c644d0219dd970aebc27e0eb044c86481307b595/cmd/clusterctl/clusterdeployer/clusterdeployer.go#L223)

This keeps the same behavior of the `Delete` method in `clusterdeployer.go` without having to call out to the `kubectl` binary. But I'm not strongly opposed to this method either. 
Apologies @juan-lee for my delay in creating my PR. I've been having a hard time creating AWS clusters using clusterctl and hence wasn't able to easily test my delete functionality.

Reply:
I also just updated the cluster functionality and not the other methods since we just have a `clusterctl delete cluster` command.